Counting Sort - English version (#106)
TheAlgorithms/Algorithms-Explanation: en/Sorting Algorithms/Couting Sort.md

# Counting Sort

#### Problem Statement

> Given an unsorted array of n elements, write a function to sort the array

Reviewer: ```suggestion
Given an unsorted array of `n` elements, write a function to sort the array.
```

#### Approach

> - Find out the maximum element (let it be max) from the given array.

Reviewer: ```suggestion
- Find out the maximum element (let's call it `max`) from the given array.
```

> - Initialize an array of length `max+1` with all elements 0 to store the count of the elements in the array.

Reviewer: ```suggestion
- Initialize an array of length `max+1` with all elements set to 0 to store the array's count.
```

> - Store the count of each element at their respective index in count array.

Reviewer: ```suggestion
- Store the count of each element at their respective index in the array's count.
```

- Store cumulative sum of the elements of the count array. It helps in placing the elements into the correct index of the sorted array.

> - Find the index of each element of the original array in the count array. This gives the cumulative count.

Reviewer: ```suggestion
- Find the index of each element of the original array in the array's count. This gives the cumulative count.
```

- Place the element at the index calculated and decrease its count by one.

#### Time Complexity

> - `O(n+k)`, best-case performance

Reviewer: ```suggestion
- `O(n+k)`: best-case performance
```

> - `O(n+k)`, average-case performance

Reviewer: ```suggestion
- `O(n+k)`: average-case performance
```

> - `O(n+k)`, worst-case performance - where k is the range of the non-negative key values.

Reviewer: ```suggestion
- `O(n+k)`: worst-case performance - where `k` is the range of the non-negative key values.
```

#### Space Complexity

> `O(n+k)`, worst-case performance

Reviewer: ```suggestion
`O(n+k)`: worst-case performance
```

#### Founder's Name

> - Harold H. Seward

Reviewer: ```suggestion
- Harold H. Seward.
```

#### Example

```
countingSort(array, size)
  max <- find largest element in array
  initialize count array with all zeros
  for j <- 0 to size
    find the total count of each unique element and
    store the count at jth index in count array
  for i <- 1 to max
    find the cumulative sum and store it in count array itself
  for j <- size down to 1
    restore the elements to array
    decrease count of each element restored by 1
```

#### Code Implementation Links

- [C](https://github.com/TheAlgorithms/C/blob/master/sorting/counting_sort.c)
- [C++](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/sorting/counting_sort.cpp)

> - [Javascript](https://github.com/TheAlgorithms/Javascript/blob/master/Sorts/CountingSort.js)

Reviewer: ```suggestion
- [JavaScript](https://github.com/TheAlgorithms/Javascript/blob/master/Sorts/CountingSort.js)
```
